lib/jsdom/living/helpers/mutation-observers.js

```diff
@@ -0,0 +1,191 @@
+"use strict";
+
+const { domSymbolTree } = require("./internal-constants");
+const reportException = require("./runtime-script-errors");
+
+const Event = require("../generated/Event");
+const idlUtils = require("../generated/utils");
+const MutationRecord = require("../generated/MutationRecord");
+
+const MUTATION_TYPE = {
+  ATTRIBUTES: "attributes",
+  CHARACTER_DATA: "characterData",
+  CHILD_LIST: "childList"
+};
+
+// https://dom.spec.whatwg.org/#queue-a-mutation-record
+function queueMutationRecord(
+  type,
+  target,
+  name,
+  namespace,
+  oldValue,
+  addedNodes,
+  removedNodes,
+  previousSibling,
+  nextSibling
+) {
+  const interestedObservers = new Map();
+
+  const nodes = domSymbolTree.ancestorsToArray(target);
+
+  for (const node of nodes) {
+    for (const registered of node._registeredObserverList) {
+      const { options, observer: mo } = registered;
+
+      if (
+        !(node !== target && options.subtree === false) &&
+        !(type === MUTATION_TYPE.ATTRIBUTES && options.attributes !== true) &&
+        !(type === MUTATION_TYPE.ATTRIBUTES && options.attributeFilter &&
+          !options.attributeFilter.some(value => value === name || value === namespace)) &&
+        !(type === MUTATION_TYPE.CHARACTER_DATA && options.characterData !== true) &&
+        !(type === MUTATION_TYPE.CHILD_LIST && options.childList === false)
+      ) {
+        if (!interestedObservers.has(mo)) {
+          interestedObservers.set(mo, null);
+        }
+
+        if (
+          (type === MUTATION_TYPE.ATTRIBUTES && options.attributeOldValue === true) ||
+          (type === MUTATION_TYPE.CHARACTER_DATA && options.characterDataOldValue === true)
+        ) {
+          interestedObservers.set(mo, oldValue);
+        }
+      }
+    }
+  }
+
+  for (const [observer, mappedOldValue] of interestedObservers.entries()) {
+    const record = MutationRecord.createImpl([], {
+      type,
+      target,
+      attributeName: name,
+      attributeNamespace: namespace,
+      oldValue: mappedOldValue,
+      addedNodes,
+      removedNodes,
+      previousSibling,
+      nextSibling
+    });
+
+    observer._recordQueue.push(record);
+  }
+
+  queueMutationObserverMicrotask();
+}
+
+// https://dom.spec.whatwg.org/#queue-a-tree-mutation-record
+function queueTreeMutationRecord(target, addedNodes, removedNodes, previousSibling, nextSibling) {
+  queueMutationRecord(
+    MUTATION_TYPE.CHILD_LIST,
+    target,
+    null,
+    null,
+    null,
+    addedNodes,
+    removedNodes,
+    previousSibling,
+    nextSibling
+  );
+}
+
+// https://dom.spec.whatwg.org/#queue-an-attribute-mutation-record
+function queueAttributeMutationRecord(target, name, namespace, oldValue) {
+  queueMutationRecord(
+    MUTATION_TYPE.ATTRIBUTES,
+    target,
+    name,
+    namespace,
+    oldValue,
+    [],
+    [],
+    null,
+    null
+  );
+}
+
+// https://dom.spec.whatwg.org/#mutation-observer-compound-microtask-queued-flag
+// TODO: Add this to the browsing context
```

Comment:
Oof. That'd be a big ask.

We have something for storage... Each window has a `_currentOriginData`. That's not quite the same, as you can be in the same origin, but not in the same UORSOBCs, if you are two independent windows with no parent/child or opener relationship.

Let's see how far we can get with this as a global. It might mess things up, but we'll see...

Reply:
Alternately we could put this on windows, although that won't work for elements created in documents without a browsing context (i.e. for which `._ownerDocument._defaultView` is undefined).

Reply:
I guess I am fine with globals for now. It's something we can iterate over in the future.